## 1. Import package & Dataset

- DataSource: The world Happiness Report : https://worldhappiness.report/

In [81]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

In [82]:
all_happiness_data = pd.read_csv("world-happiness-report-raw.csv")
all_happiness_data

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008        3.724               7.370           0.451   
1     Afghanistan  2009        4.402               7.540           0.552   
2     Afghanistan  2010        4.758               7.647           0.539   
3     Afghanistan  2011        3.832               7.620           0.521   
4     Afghanistan  2012        3.783               7.705           0.521   
...           ...   ...          ...                 ...             ...   
1944     Zimbabwe  2016        3.735               7.984           0.768   
1945     Zimbabwe  2017        3.638               8.016           0.754   
1946     Zimbabwe  2018        3.616               8.049           0.775   
1947     Zimbabwe  2019        2.694               7.950           0.759   
1948     Zimbabwe  2020        3.160               7.829           0.717   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                                50.80                         0.718   
1                                51.20                         0.679   
2                                51.60                         0.600   
3                                51.92                         0.496   
4                                52.24                         0.531   
...                                ...                           ...   
1944                             54.40                         0.733   
1945                             55.00                         0.753   
1946                             55.60                         0.763   
1947                             56.20                         0.632   
1948                             56.80                         0.643   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  
0          0.168                      0.882            0.518            0.258  
1          0.190                      0.850            0.584            0.237  
2          0.121                      0.707            0.618            0.275  
3          0.162                      0.731            0.611            0.267  
4          0.236                      0.776            0.710            0.268  
...          ...                        ...              ...              ...  
1944      -0.095                      0.724            0.738            0.209  
1945      -0.098                      0.751            0.806            0.224  
1946      -0.068                      0.844            0.710            0.212  
1947      -0.064                      0.831            0.716            0.235  
1948      -0.009                      0.789            0.703            0.346  

[1949 rows x 11 columns]

## 2. Get data ready

- Clean Data

In [83]:
all_happiness_data.count()

Country name                        1949
year                                1949
Life Ladder                         1949
Log GDP per capita                  1913
Social support                      1936
Healthy life expectancy at birth    1894
Freedom to make life choices        1917
Generosity                          1860
Perceptions of corruption           1839
Positive affect                     1927
Negative affect                     1933
dtype: int64

### 2-1 Overview

- How Many Country Are there?

In [84]:
n = len(pd.unique(all_happiness_data['Country name']))
print("Number of unique Countries :", n)

Number of unique Countries : 166


- How Many Years Are there?

In [85]:
n = len(pd.unique(all_happiness_data['year']))
print("Number of unique year :", n)

Number of unique year : 16


- We have total of 1949 rows
- However 166(Countries) * 16(year) = 2656
- Means we have some missing years within some countries, let's identify them.

- Is the year for the countries align??

In [86]:
# sort dataset by country & year
all_happiness_data.sort_values(by=['Country name', 'year'])
all_happiness_data.head()

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008        3.724               7.370           0.451   
1  Afghanistan  2009        4.402               7.540           0.552   
2  Afghanistan  2010        4.758               7.647           0.539   
3  Afghanistan  2011        3.832               7.620           0.521   
4  Afghanistan  2012        3.783               7.705           0.521   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                             50.80                         0.718       0.168   
1                             51.20                         0.679       0.190   
2                             51.60                         0.600       0.121   
3                             51.92                         0.496       0.162   
4                             52.24                         0.531       0.236   

   Perceptions of corruption  Positive affect  Negative affect  
0                      0.882            0.518            0.258  
1                      0.850            0.584            0.237  
2                      0.707            0.618            0.275  
3                      0.731            0.611            0.267  
4                      0.776            0.710            0.268

In [87]:
all_happiness_data['year'].value_counts().sort_index()

2005     27
2006     89
2007    102
2008    110
2009    114
2010    124
2011    146
2012    142
2013    137
2014    145
2015    143
2016    142
2017    147
2018    142
2019    144
2020     95
Name: year, dtype: int64

- As most of the county have data from 2007 to 2020. I want to elimiate the data from 2005 & 2006

In [88]:
all_happiness_data = all_happiness_data[all_happiness_data['year'] > 2006]

In [89]:
# let's check
all_happiness_data['year'].value_counts().sort_index()

2007    102
2008    110
2009    114
2010    124
2011    146
2012    142
2013    137
2014    145
2015    143
2016    142
2017    147
2018    142
2019    144
2020     95
Name: year, dtype: int64

### 2-2 Deal with missing values

In [90]:
# Check the null count
print('Count of null values by column:')
all_happiness_data.isna().sum()

Count of null values by column:


Country name                          0
year                                  0
Life Ladder                           0
Log GDP per capita                   35
Social support                       13
Healthy life expectancy at birth     54
Freedom to make life choices         27
Generosity                           55
Perceptions of corruption           102
Positive affect                      21
Negative affect                      14
dtype: int64

### 2-2-1 Set up approach for each column to deal with missing values

- As all of our missing column is numerical, fill with the average within the country will be fair

In [91]:
mean_df = all_happiness_data.groupby('Country name', as_index = False).mean()

In [92]:
for index, row in all_happiness_data.iterrows():

        if(pd.isnull(row['Log GDP per capita'])):
            all_happiness_data.at[index , 'Log GDP per capita'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Log GDP per capita']
        if(pd.isnull(row['Social support'])):
            all_happiness_data.at[index , 'Social support'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Social support']
        if(pd.isnull(row['Healthy life expectancy at birth'])):
            all_happiness_data.at[index , 'Healthy life expectancy at birth'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Healthy life expectancy at birth']
        if(pd.isnull(row['Freedom to make life choices'])):    
            all_happiness_data.at[index , 'Freedom to make life choices'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Freedom to make life choices']
        if(pd.isnull(row['Generosity'])):    
            all_happiness_data.at[index , 'Generosity'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Generosity']
        if(pd.isnull(row['Perceptions of corruption'])):    
            all_happiness_data.at[index , 'Perceptions of corruption'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Perceptions of corruption']
        if(pd.isnull(row['Positive affect'])):    
            all_happiness_data.at[index , 'Positive affect'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Positive affect']
        if(pd.isnull(row['Negative affect'])):    
            all_happiness_data.at[index , 'Negative affect'] = mean_df.loc[mean_df['Country name'] == row['Country name']]['Negative affect']

In [93]:
# Check the null count
print('Count of null values by column:')
all_happiness_data.isna().sum()

Count of null values by column:


Country name                         0
year                                 0
Life Ladder                          0
Log GDP per capita                  18
Social support                       1
Healthy life expectancy at birth    35
Freedom to make life choices         0
Generosity                          18
Perceptions of corruption           26
Positive affect                      2
Negative affect                      1
dtype: int64

- There are still some missing data, and I will like to fill them with average of the column as I don't want to delet them to loss any more data

In [94]:
all_happiness_data.fillna(all_happiness_data.mean(), inplace=True)

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.
/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [95]:
# Check the null count
print('Count of null values by column:')
all_happiness_data.isna().sum()

Count of null values by column:


Country name                        0
year                                0
Life Ladder                         0
Log GDP per capita                  0
Social support                      0
Healthy life expectancy at birth    0
Freedom to make life choices        0
Generosity                          0
Perceptions of corruption           0
Positive affect                     0
Negative affect                     0
dtype: int64

### 2-3 Add Continent Column

- As I will also like to analysis countries based on Continent, I will like to add this column to the all_happiness_data
- Find country to continent on Kaggle: https://www.kaggle.com/statchaitya/country-to-continent

In [96]:
continent_data = pd.read_csv("countryContinent.csv",encoding='ISO-8859-1')
continent_data

country code_2 code_3  country_code     iso_3166_2 continent  \
0          Afghanistan     AF    AFG             4  ISO 3166-2:AF      Asia   
1       Ã
land Islands     AX    ALA           248  ISO 3166-2:AX    Europe   
2              Albania     AL    ALB             8  ISO 3166-2:AL    Europe   
3              Algeria     DZ    DZA            12  ISO 3166-2:DZ    Africa   
4       American Samoa     AS    ASM            16  ISO 3166-2:AS   Oceania   
..                 ...    ...    ...           ...            ...       ...   
244  Wallis and Futuna     WF    WLF           876  ISO 3166-2:WF   Oceania   
245     Western Sahara     EH    ESH           732  ISO 3166-2:EH    Africa   
246              Yemen     YE    YEM           887  ISO 3166-2:YE      Asia   
247             Zambia     ZM    ZMB           894  ISO 3166-2:ZM    Africa   
248           Zimbabwe     ZW    ZWE           716  ISO 3166-2:ZW    Africa   

          sub_region  region_code  sub_region_code  
0      Southern Asia        142.0             34.0  
1    Northern Europe        150.0            154.0  
2    Southern Europe        150.0             39.0  
3    Northern Africa          2.0             15.0  
4          Polynesia          9.0             61.0  
..               ...          ...              ...  
244        Polynesia          9.0             61.0  
245  Northern Africa          2.0             15.0  
246     Western Asia        142.0            145.0  
247   Eastern Africa          2.0             14.0  
248   Eastern Africa          2.0             14.0  

[249 rows x 9 columns]

### 2-3-1 Clean continent_data as I only need country name & continent & sub_region

In [97]:
continent_data = continent_data[['country', 'continent', 'sub_region']]
continent_data.rename(columns={'country':'Country name'}, inplace=True)
continent_data

/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:5047: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Country name continent       sub_region
0          Afghanistan      Asia    Southern Asia
1       Ã
land Islands    Europe  Northern Europe
2              Albania    Europe  Southern Europe
3              Algeria    Africa  Northern Africa
4       American Samoa   Oceania        Polynesia
..                 ...       ...              ...
244  Wallis and Futuna   Oceania        Polynesia
245     Western Sahara    Africa  Northern Africa
246              Yemen      Asia     Western Asia
247             Zambia    Africa   Eastern Africa
248           Zimbabwe    Africa   Eastern Africa

[249 rows x 3 columns]

### 2-3-2 Left join all_happiness_data & continent_data

In [98]:
all_happiness_data = pd.merge(all_happiness_data, continent_data, on="Country name", how='left')

In [99]:
all_happiness_data

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008        3.724               7.370           0.451   
1     Afghanistan  2009        4.402               7.540           0.552   
2     Afghanistan  2010        4.758               7.647           0.539   
3     Afghanistan  2011        3.832               7.620           0.521   
4     Afghanistan  2012        3.783               7.705           0.521   
...           ...   ...          ...                 ...             ...   
1828     Zimbabwe  2016        3.735               7.984           0.768   
1829     Zimbabwe  2017        3.638               8.016           0.754   
1830     Zimbabwe  2018        3.616               8.049           0.775   
1831     Zimbabwe  2019        2.694               7.950           0.759   
1832     Zimbabwe  2020        3.160               7.829           0.717   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                                50.80                         0.718   
1                                51.20                         0.679   
2                                51.60                         0.600   
3                                51.92                         0.496   
4                                52.24                         0.531   
...                                ...                           ...   
1828                             54.40                         0.733   
1829                             55.00                         0.753   
1830                             55.60                         0.763   
1831                             56.20                         0.632   
1832                             56.80                         0.643   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  \
0          0.168                      0.882            0.518            0.258   
1          0.190                      0.850            0.584            0.237   
2          0.121                      0.707            0.618            0.275   
3          0.162                      0.731            0.611            0.267   
4          0.236                      0.776            0.710            0.268   
...          ...                        ...              ...              ...   
1828      -0.095                      0.724            0.738            0.209   
1829      -0.098                      0.751            0.806            0.224   
1830      -0.068                      0.844            0.710            0.212   
1831      -0.064                      0.831            0.716            0.235   
1832      -0.009                      0.789            0.703            0.346   

     continent      sub_region  
0         Asia   Southern Asia  
1         Asia   Southern Asia  
2         Asia   Southern Asia  
3         Asia   Southern Asia  
4         Asia   Southern Asia  
...        ...             ...  
1828    Africa  Eastern Africa  
1829    Africa  Eastern Africa  
1830    Africa  Eastern Africa  
1831    Africa  Eastern Africa  
1832    Africa  Eastern Africa  

[1833 rows x 13 columns]

In [100]:
n = len(pd.unique(all_happiness_data['Country name']))
print("Number of unique Countries :", n)

Number of unique Countries : 165


In [101]:
# Check the null count
print('Count of null values by column:')
all_happiness_data.isna().sum()

Count of null values by column:


Country name                         0
year                                 0
Life Ladder                          0
Log GDP per capita                   0
Social support                       0
Healthy life expectancy at birth     0
Freedom to make life choices         0
Generosity                           0
Perceptions of corruption            0
Positive affect                      0
Negative affect                      0
continent                           60
sub_region                          60
dtype: int64

### 2-3-3 Check country with empty continent

In [102]:
all_happiness_data[all_happiness_data['continent'].isnull()]

Country name  year  Life Ladder  Log GDP per capita  \
402                    Cyprus  2009        6.833           10.558000   
403                    Cyprus  2010        6.387           10.551000   
404                    Cyprus  2011        6.690           10.530000   
405                    Cyprus  2012        6.181           10.479000   
406                    Cyprus  2013        5.439           10.414000   
407                    Cyprus  2014        5.627           10.406000   
408                    Cyprus  2015        5.439           10.445000   
409                    Cyprus  2016        5.795           10.506000   
410                    Cyprus  2017        6.062           10.539000   
411                    Cyprus  2018        6.276           10.567000   
412                    Cyprus  2019        6.137           10.585000   
413                    Cyprus  2020        6.260           10.507273   
782               Ivory Coast  2009        4.197            8.209000   
783               Ivory Coast  2013        3.739            8.274000   
784               Ivory Coast  2014        3.570            8.334000   
785               Ivory Coast  2015        4.445            8.393000   
786               Ivory Coast  2016        4.543            8.437000   
787               Ivory Coast  2017        5.038            8.483000   
788               Ivory Coast  2018        5.268            8.523000   
789               Ivory Coast  2019        5.392            8.564000   
790               Ivory Coast  2020        5.257            8.565000   
852                    Kosovo  2007        5.104            8.928000   
853                    Kosovo  2008        5.522            8.981000   
854                    Kosovo  2009        5.891            9.008000   
855                    Kosovo  2010        5.177            9.033000   
856                    Kosovo  2011        4.860            9.067000   
857                    Kosovo  2012        5.640            9.086000   
858                    Kosovo  2013        6.126            9.113000   
859                    Kosovo  2014        5.000            9.129000   
860                    Kosovo  2015        5.077            9.182000   
861                    Kosovo  2016        5.759            9.228000   
862                    Kosovo  2017        6.149            9.262000   
863                    Kosovo  2018        6.392            9.296000   
864                    Kosovo  2019        6.425            9.339000   
865                    Kosovo  2020        6.294            9.127077   
890                      Laos  2007        5.364            8.307000   
891                      Laos  2008        5.044            8.366000   
892                      Laos  2011        4.704            8.548000   
893                      Laos  2012        4.876            8.611000   
894                      Laos  2017        4.623            8.890000   
895                      Laos  2018        4.859            8.935000   
896                      Laos  2019        5.197            8.965000   
897                      Laos  2020        5.284            8.960000   
1250  Palestinian Territories  2007        4.151            8.218000   
1251  Palestinian Territories  2008        4.386            8.276000   
1252  Palestinian Territories  2009        4.470            8.329000   
1253  Palestinian Territories  2010        4.703            8.383000   
1254  Palestinian Territories  2011        4.751            8.474000   
1255  Palestinian Territories  2012        4.647            8.531000   
1256  Palestinian Territories  2013        4.844            8.489000   
1257  Palestinian Territories  2014        4.722            8.457000   
1258  Palestinian Territories  2015        4.695            8.480000   
1259  Palestinian Territories  2016        4.907            8.498000   
1260  Palestinian Territories  2017        4.628            8.485000   
1261  Palestinian Territories  2018        4.554         

## 3 Save csv

In [104]:
all_happiness_data.to_csv("world-happiness-report-new.csv", index=False)